plugin/caffe/caffe_loss-inl.h

```diff
@@ -0,0 +1,266 @@
+    grad_scale_ = (Dtype)param_.grad_scale;
+
+    InitCaffeBlobs<Dtype>(bot_, param_.in_num);
+    InitCaffeBlobs<Dtype>(top_, param_.out_num);
+    flags_.resize(param_.in_num);
+  }
+
+  ~CaffeLoss() {
+    DelCaffeBlobs(bot_, param_.in_num);
+    DelCaffeBlobs(top_, param_.out_num);
+  }
+
+  virtual void Forward(const OpContext &ctx,
+                       const std::vector<TBlob> &in_data,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    // Set mode before forward
+    ::mxnet::CaffeMode::SetMode<xpu>();
+    using ::caffe::Blob;
+    using std::vector;
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    for (index_t i = 0; i < req.size(); ++i)
+      CHECK_EQ(req[i], kWriteTo);
+
+    CHECK_EQ(in_data.size(), param_.in_num);
+    CHECK_EQ(out_data.size(), param_.out_num);
+
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+#if defined(__CUDACC__)
+    // TODO(Haoran): when need cublas handle in stream?
+    CHECK_EQ(s->blas_handle_ownership_, Stream<xpu>::OwnHandle)
+          << "Must init CuBLAS handle in stream";
+#endif  // __CUDACC__
+
+    TBlob2CaffeBlob<xpu, Dtype>(caffememtype::Data, bot_.begin(), in_data.begin(), param_.in_num);
+    TBlob2CaffeBlob<xpu, Dtype>(caffememtype::Data, top_.begin(), out_data.begin(), param_.out_num);
+
+    CaffeOpSetup();
+    caffeOp_->Forward(bot_, top_);
+  }
+
+  // Set up caffe op with real data
+  void CaffeOpSetup() {
+    if (!setup_ ) {
+      setup_ = true;
+      caffeOp_->SetUp(bot_, top_);
+    }
+  }
+
+  virtual void Backward(const OpContext &ctx,
+                        const std::vector<TBlob> &out_grad,
+                        const std::vector<TBlob> &in_data,
+                        const std::vector<TBlob> &out_data,
+                        const std::vector<OpReqType> &req,
+                        const std::vector<TBlob> &in_grad,
+                        const std::vector<TBlob> &aux_args) {
+    // Set mode before backward
+    ::mxnet::CaffeMode::SetMode<xpu>();
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    CHECK_EQ(out_grad.size(), param_.out_num);
+    for (index_t i = 0; i < param_.in_num; ++i)
+      CHECK(req[i] != kAddTo) << "caffe doesn't accm diff on bottom data";
+    CHECK(in_data.size() == param_.in_num);
+
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+#if defined(__CUDACC__)
+    // TODO(Haoran): when need cublas handle in stream?
+    CHECK_EQ(s->blas_handle_ownership_, Stream<xpu>::OwnHandle)
+          << "Must init CuBLAS handle in stream";
+#endif  // __CUDACC__
+
+    TBlob2CaffeBlob<xpu, Dtype>(caffememtype::Grad, bot_.begin(), in_grad.begin(), param_.in_num);
+    // Pass grad scale to caffe blob
+    top_[0]->set_cpu_diff(&grad_scale_);
+
+    // Set BP flag 
+    for (index_t i = 0; i < param_.in_num; ++i)
+      flags_[i] = req[i] != kNullOp;
```

Comment:
caffe doesn't support addto right? should check here.
Also I forgot how caffe handles inplace. Probably need to check too